### PR TITLE
feat(logging): Add Cloudwatch Logging Plugin

### PR DESCRIPTION
**Pull request by sdhuka**
- [ ] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
* Add cloudwatch logging plugin
*How did you test these changes?*
* Unit & Manual testing using pentest app.

*Documentation update required?*
- [ ] No
- [X] Yes (Will be added later)

*General Checklist*
- [X] Added Unit Tests
- [ ] Added Integration Tests
- [X] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-logging-cloudwatch/gradle.properties [hunk 1]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=aws-logging-cloudwatch
+POM_NAME=Amplify Framework for Android - Logging
+POM_DESCRIPTION=Amplify Framework for Android - Logging Plugin for AWS CloudWatch
```
Comment (div5yesh) [inline aws-logging-cloudwatch/gradle.properties:3]: Amazon CloudWatch, not AWS.

File: aws-logging-cloudwatch/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/AndroidManifest.xml:2]: missing license header.

File: build.gradle.kts [hunk 3]
```diff
@@ -101,7 +101,7 @@ subprojects {
 
     tasks.withType<Test>().configureEach {
         retry {
-            maxRetries.set(9)
+            maxRetries.set(1)
```
Comment (div5yesh) [inline build.gradle.kts:104]: Is this change intentional?

File: settings.gradle.kts [hunk 4]
```diff
@@ -139,6 +141,7 @@ dependencyResolutionManagement {
             library("gson", "com.google.code.gson:gson:2.8.9")
             library("tensorflow", "org.tensorflow:tensorflow-lite:2.0.0")
             library("uuidgen", "com.fasterxml.uuid:java-uuid-generator:4.0.1")
+            library("sqlcipher", "net.zetetic:android-database-sqlcipher:4.5.3")
```
Comment (div5yesh) [inline settings.gradle.kts:144]: Please follow up: license and security.

File: core/src/test/java/com/amplifyframework/logging/FakeLoggingPlugin.java [hunk 5]
```diff
@@ -20,6 +20,7 @@
 import androidx.annotation.Nullable;
 
 import com.amplifyframework.core.BuildConfig;
+import com.amplifyframework.core.category.CategoryType;
```
Comment (div5yesh) [inline core/src/test/java/com/amplifyframework/logging/FakeLoggingPlugin.java:23]: update the license header

File: core/src/main/java/com/amplifyframework/core/category/Category.java [hunk 6]
```diff
@@ -86,9 +86,11 @@ public synchronized void configure(
             state.set(State.CONFIGURING);
             try {
                 for (P plugin : getPlugins()) {
-                    String pluginKey = plugin.getPluginKey();
-                    JSONObject pluginConfig = configuration.getPluginConfig(pluginKey);
-                    plugin.configure(pluginConfig != null ? pluginConfig : new JSONObject(), context);
+                    if (configureFromDefaultConfigFile()) {
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/core/category/Category.java:89]: why do we need this?

File: core/src/main/java/com/amplifyframework/logging/LoggingCategoryBehavior.java [hunk 7]
```diff
@@ -18,6 +18,8 @@
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
+import com.amplifyframework.core.category.CategoryType;
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/logging/LoggingCategoryBehavior.java:21]: update license header

File: core/src/main/java/com/amplifyframework/logging/LoggingCategory.java [hunk 8]
```diff
@@ -74,9 +81,78 @@ public Logger forNamespace(@Nullable String namespace) {
...
+        Set<LoggingPlugin<?>> loggingPlugins = new HashSet<>(getPlugins());
+        loggingPlugins.add(defaultPlugin);
+        for (LoggingPlugin<?> plugin : loggingPlugins) {
+            plugin.configure(loggingConfiguration, context);
+        }
+    }
+
+    private JSONObject readConfigFile(Context context) {
+        try {
+            String configName = "amplifyconfiguration_logging";
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/logging/LoggingCategory.java:151]: Are we going to have two config files?

**Comment by div5yesh** [comment]
Should the target branch be `main`?

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogger.kt [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        val event = CloudWatchLogEvent(System.currentTimeMillis(), "verbose/$namespace: $message")
+        persistEvent(event)
+    }
+
+    private fun persistEvent(event: CloudWatchLogEvent) {
+        coroutineScope.launch {
+            awsCloudWatchLoggingPlugin.cloudWatchLogManager?.saveLogEvent(event)?.also {
+                flushLogsQueue()
+            } ?: kotlin.run {
```
Comment (mattcreaser) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogger.kt:117]: This run block doesn't seem to be doing anything

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogger.kt [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            return
+        }
+        val event = CloudWatchLogEvent(
+            System.currentTimeMillis(),
+            "warn/$namespace: $message, issue: ${Log.getStackTraceString(issue)}",
+        )
+        persistEvent(event)
+    }
+
+    override fun info(message: String?) {
```
Comment (mattcreaser) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogger.kt:89]: These functions implementations are repeated, can make a helper

```kotlin
override fun info(message: String) = log(LogLevel.INFO, message)

private fun log(level: LogLevel, message: String?) {
  if (shouldNotLogMessage(level)) {
    return
  }
  val event = CloudWatchLogEvent(System.currentTimeMillis(), "${level.name.lowercase()}/$namespace: $message")
  persistEvent(event)
}

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogManager.kt [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        onSignIn()
+    }
+
+    suspend fun saveLogEvent(event: CloudWatchLogEvent) {
+        withContext(coroutineDispatcher) {
+            try {
+                cloudWatchLoggingDatabase.saveLogEvent(event)
+                if (isCacheFull()) {
+                    syncLogEventsWithCloudwatch()
+                } else {
```
Comment (mattcreaser) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogManager.kt:73]: empty else

**Review by mattcreaser — COMMENTED** [review]
There are many warnings for things like extra trailing commas at the end of parameter lists - I stopped commenting on them all so please check. KtLint plugin will help 😄 

I'm also concerned about the use of TimerTask, as it won't be reliable during process death or when the OS implements various battery saving measures. The two things we are using for it here are explicitly called out as the correct [use cases for work manager](https://developer.android.com/topic/libraries/architecture/workmanager#reliable).

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/CloudWatchLogEvent.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.logging.cloudwatch.models
+
+internal data class CloudWatchLogEvent(val timestamp: Long, val message: String, val id: Int = 0)
```
Comment (tylerjroach) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/CloudWatchLogEvent.kt:17]: I think id should be a long here according to Sqlite integer primary keys

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    private val awsCloudWatchLoggingPluginImplementation = AWSCloudWatchLoggingPluginImplementation(
+        loggingConstraintsResolver,
+        awsCloudWatchLoggingPluginConfig,
+    )
+    private lateinit var cloudWatchLogsClient: CloudWatchLogsClient
+    private val logger = Amplify.Logging.logger(CategoryType.LOGGING, this::class.java.simpleName)
+
+    companion object {
+        private const val CONFIG_FILENAME = "amplifyconfiguration_logging"
+        internal const val SHARED_PREFERENCE_FILENAME = "com.amplify.logging.a3fa4188-0ac5-11ee-be56-0242ac120002"
```
Comment (tylerjroach) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:56]: whats the uuid at the end of this?

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/CloudWatchLogEvent.kt (hunk 12 above)
Comment (tylerjroach) [reply aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/CloudWatchLogEvent.kt:17]: I also don't like the double usage here making id a default value of 0, but also pulling a CloudWatchLogEvent from db that has a real id. These should be split up.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+    public fun flushLogs(
+        onSuccess: Action,
+        onError: Consumer<AmplifyException>,
+    ) {
+        awsCloudWatchLoggingPluginImplementation.flushLogs(onSuccess, onError)
+    }
+
+    override fun getPluginKey(): String {
+        return "awsCloudWatchLoggingPlugin"
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:88]: Please move this to constants above.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+                        ),
+                    )
+                }
+            }
+            val cloudWatchLogManager =
+                CloudWatchLogManager(context, awsLoggingConfig, cloudWatchLogsClient, loggingConstraintsResolver)
+            awsCloudWatchLoggingPluginImplementation.cloudWatchLogManager = cloudWatchLogManager
+            awsCloudWatchLoggingPluginImplementation.configure(awsLoggingConfig)
+        } catch (exception: Exception) {
+            logger.error("failed to configure plugin", exception)
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:115]: This should throw an AmplifyException: https://github.com/aws-amplify/amplify-android/blob/main/aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt#L146

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    override fun getEscapeHatch(): CloudWatchLogsClient {
+        return cloudWatchLogsClient
+    }
+
+    override fun getVersion(): String = BuildConfig.VERSION_NAME
+
+    @OptIn(ExperimentalSerializationApi::class)
+    private fun getConfigFromFile(pluginConfiguration: JSONObject?): AWSCloudWatchLoggingPluginConfiguration {
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:126]: nit: Please move this to AWSCloudWatchLoggingPluginConfiguration.

File: core/src/main/java/com/amplifyframework/logging/LoggingCategory.java [hunk 17]
```diff
@@ -74,9 +81,78 @@ public Logger forNamespace(@Nullable String namespace) {
...
+        for (LoggingPlugin<?> plugin : loggingPlugins) {
+            delegates.add(plugin.logger(namespace));
+        }
+        return new BroadcastLogger(delegates);
+    }
+
+    @NonNull
+    @Override
+    public Logger logger(@NonNull CategoryType categoryType, @NonNull String namespace) {
+        Set<LoggingPlugin<?>> loggingPlugins = new HashSet<>(getPlugins());
```
[the remark this answers is not in the document]
Comment (div5yesh) [reply core/src/main/java/com/amplifyframework/logging/LoggingCategory.java:99]: +1

File: core/src/main/java/com/amplifyframework/logging/AndroidLogger.java [hunk 18]
```diff
@@ -23,6 +23,7 @@
 import java.util.Objects;
 
 final class AndroidLogger implements Logger {
+    private static boolean isEnabled = true;
```
Comment (div5yesh) [inline core/src/main/java/com/amplifyframework/logging/AndroidLogger.java:26]: I think this should be part of the Logger interface?

File: core/src/main/java/com/amplifyframework/logging/AndroidLoggingPlugin.java [hunk 19]
```diff
@@ -53,11 +54,34 @@ public AndroidLoggingPlugin(@NonNull LogLevel defaultLoggerThreshold) {
...
+    @NonNull
+    @Override
+    public Logger logger(@NonNull String namespace) {
+        return new AndroidLogger(namespace, defaultLoggerThreshold);
+    }
+
+    @NonNull
+    @Override
+    public Logger logger(@NonNull CategoryType categoryType, @NonNull String namespace) {
+        return new AndroidLogger(namespace, defaultLoggerThreshold);
```
[the remark this answers is not in the document]
Comment (div5yesh) [reply core/src/main/java/com/amplifyframework/logging/AndroidLoggingPlugin.java:72]: I think categoryType should be a part of the namespace since AndroidLogger does not have a categoryType param.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogManager.kt [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+            inputLogEventsIdToBeDeleted.add(cloudWatchEvent.id)
+            iterator.remove()
+        }
+        return Pair(inputLogEvents, inputLogEventsIdToBeDeleted)
+    }
+
+    private fun clearCache() {
+        coroutineScope.launch {
+            cloudWatchLoggingDatabase.clearDatabase()
+            context.getSharedPreferences(
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CloudWatchLogManager.kt:210]: Please extract Shared Preferences operations into a separate type.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+import kotlinx.serialization.json.Json
+
+/**
+ * Constraints to control the log level per categoryType or user
+ */
+@Serializable
+data class LoggingConstraints @JvmOverloads constructor(
+    val defaultLogLevel: LogLevel = LogLevel.ERROR,
+    val categoryLogLevel: Map<CategoryType, LogLevel> = emptyMap(),
+    val userLogLevel: Map<String, UserLogLevel> = emptyMap(),
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt:32]: It seems the userLogLevel overrides the root level categoryLogLevel, so why not design the types in such a way? 

Something like:
```
UserLoggingConstraints: LoggingConstraints {
    override val defaultLogLevel: LogLevel,
    override val categoryLogLevel: Map<CategoryType, LogLevel>
}
```

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+) {
+    companion object {
+        @OptIn(ExperimentalSerializationApi::class)
+        private val json = Json {
+            encodeDefaults = true
+            explicitNulls = false
+            ignoreUnknownKeys = true
+        }
+        fun fromString(jsonString: String): LoggingConstraints {
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt:41]: `internal`?

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+            return json.decodeFromString<LoggingConstraints>(jsonString)
+        }
+
+        fun toJsonString(loggingConstraints: LoggingConstraints): String {
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/models/LoggingConstraints.kt:45]: `internal`?

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/db/CloudWatchLoggingDatabase.kt [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+        )
+    }
+
+    private fun getInstallationIdentifier(context: Context): String {
+        val identifierFile = File(context.noBackupFilesDir, "awscloudwatchloggingdb.installationIdentifier")
+        val previousIdentifier = getExistingInstallationIdentifier(identifierFile)
+        return previousIdentifier ?: createInstallationIdentifier(identifierFile)
+    }
+
+    private fun getExistingInstallationIdentifier(identifierFile: File): String? {
```
[the remark this answers is not in the document]
Comment (div5yesh) [reply aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/db/CloudWatchLoggingDatabase.kt:156]: Please use the common Credential Store instead.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+    internal fun flushLogs(
+        onSuccess: Action,
+        onError: Consumer<AmplifyException>,
+    ) {
+        coroutineScope.launch {
+            try {
+                cloudWatchLogManager?.syncLogEventsWithCloudwatch()
+                onSuccess.call()
+            } catch (exception: Exception) {
+                onError.accept(AmplifyException("Failed to flush logs", exception.cause, "Please try again"))
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt:108]: Plugins should return `<Category>Exceptions` and not AmplifyExceptions (except from `configure()`).

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+    private val logger = Amplify.Logging.logger(CategoryType.LOGGING, this::class.java.simpleName)
+
+    companion object {
+        private const val CONFIG_FILENAME = "amplifyconfiguration_logging"
+        internal const val SHARED_PREFERENCE_FILENAME = "com.amplify.logging.a3fa4188-0ac5-11ee-be56-0242ac120002"
+    }
+
+    @Deprecated("Deprecated in Java")
+    override fun forNamespace(namespace: String?): Logger {
+        return awsCloudWatchLoggingPluginImplementation.forNamespace(namespace)
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:61]: This should directly call `logger(namespace)` and we can remove `forNamespace` from AWSCloudWatchLoggingPluginImplementation.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+            if (event.name == AuthChannelEventName.SIGNED_IN.name) {
+                cloudWatchLogManager?.onSignIn()
+            }
+            if (event.name == AuthChannelEventName.SIGNED_OUT.name) {
+                cloudWatchLogManager?.onSignOut()
+            }
+        }
+    }
+
+    internal fun forNamespace(namespace: String?): Logger {
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt:62]: Not required, see comment above in AWSCloudWatchLoggingPlugin.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+        val resolvedNameSpace = namespace ?: defaultNameSpace
+        return logger(resolvedNameSpace)
+    }
+
+    internal fun logger(namespace: String): Logger {
+        return CloudWatchLogger(
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt:68]: How does cloudWatchLogManager know about `CloudWatchLogger`? Can you point me to code where these two interact?

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CustomCognitoCredentialsProvider.kt [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+package com.amplifyframework.logging.cloudwatch
+
+import com.amplifyframework.auth.AuthUser
+import com.amplifyframework.auth.CognitoCredentialsProvider
+import com.amplifyframework.core.Amplify
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+internal class CustomCognitoCredentialsProvider() : CognitoCredentialsProvider() {
```
Comment (div5yesh) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/CustomCognitoCredentialsProvider.kt:24]: Why not add `getCurrentUser` to `CognitoCredentialsProvider` instead?

File: core/src/main/java/com/amplifyframework/devmenu/PersistentLogStoragePlugin.java [hunk 30]
```diff
@@ -50,18 +51,41 @@ public PersistentLogStoragePlugin() {
...
+            PersistentLogger newLogger = new PersistentLogger(namespace);
+            loggers.put(namespace, newLogger);
             return newLogger;
         }
     }
 
+    @NonNull
+    @Override
+    public Logger logger(@NonNull CategoryType categoryType, @NonNull String namespace) {
+        return logger(namespace);
```
[the remark this answers is not in the document]
Comment (div5yesh) [reply core/src/main/java/com/amplifyframework/devmenu/PersistentLogStoragePlugin.java:76]: Same as previous comment, I think categoryType should be a part of namespace.

File: core/src/test/java/com/amplifyframework/logging/FakeLoggingPlugin.java (hunk 5 above)
Comment (sdhuka) [reply core/src/test/java/com/amplifyframework/logging/FakeLoggingPlugin.java:23]: confirmed this is pre-approved by OSS.

File: core/src/main/java/com/amplifyframework/core/category/Category.java (hunk 6 above)
Comment (sdhuka) [reply core/src/main/java/com/amplifyframework/core/category/Category.java:89]: logging config is not read from amplifyconfiguration.json so this change is added to prevent calling plugin `configure()` for categories where configuration is read from default config file.

File: core/src/main/java/com/amplifyframework/logging/LoggingCategoryBehavior.java (hunk 7 above)
Comment (sdhuka) [reply core/src/main/java/com/amplifyframework/logging/LoggingCategoryBehavior.java:21]: this file is updated, not added new.

File: core/src/main/java/com/amplifyframework/logging/LoggingCategory.java [hunk 31]
```diff
@@ -74,9 +81,78 @@ public Logger forNamespace(@Nullable String namespace) {
...
+        loggingPlugins.add(defaultPlugin);
+        for (LoggingPlugin<?> plugin : loggingPlugins) {
+            plugin.configure(loggingConfiguration, context);
+        }
+    }
+
+    private JSONObject readConfigFile(Context context) {
+        try {
+            String configName = "amplifyconfiguration_logging";
+            int resourceId = Resources.getRawResourceId(context, configName);
```
[the remark this answers is not in the document]
Comment (sdhuka) [reply core/src/main/java/com/amplifyframework/logging/LoggingCategory.java:152]: config file is added in apps raw folder which is not available to reference directly from library.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt (hunk 13 above)
Comment (sdhuka) [reply aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPlugin.kt:56]: It is added for uniqueness.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt (hunk 27 above)
Comment (sdhuka) [reply aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt:62]: it resolves to default incase namespace is null.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt (hunk 28 above)
Comment (sdhuka) [reply aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt:68]: CloudWatchLogger has pluginImplementation.

File: core/src/main/java/com/amplifyframework/logging/AndroidLoggingPlugin.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (sdhuka) [reply core/src/main/java/com/amplifyframework/logging/AndroidLoggingPlugin.java:72]: categotyType filter are only supported for cloudwatch logger. This filter might be added in future for other loggers.

File: core/src/main/java/com/amplifyframework/logging/LoggingEventName.kt [hunk 32]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.logging
+
+/**
+ * An enumeration of the names of events relating the {@link LoggingCategory},
```
Comment (mattcreaser) [inline core/src/main/java/com/amplifyframework/logging/LoggingEventName.kt:18]: In kdoc these links should just be brackets e.g. `[LoggingCategory]`, see [KDoc syntax](https://kotlinlang.org/docs/kotlin-doc.html#links-to-elements) for details.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/db/CloudWatchLoggingDatabase.kt [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+import kotlinx.coroutines.withContext
+import net.sqlcipher.database.SQLiteQueryBuilder
+
+internal class CloudWatchLoggingDatabase(
+    private val context: Context,
+    private val coroutineDispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+    private val logEvents = 10
+    private val logEventsId = 20
+    private val passphraseKey = "passphrase"
```
Comment (tylerjroach) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/db/CloudWatchLoggingDatabase.kt:36]: These should all be const in companion object.

File: aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/DefaultRemoteLoggingConstraintProvider.kt [hunk 34]
```diff
@@ -0,0 +1,118 @@
...
+import okhttp3.OkHttpClient
+import okhttp3.Request
+
+/**
+ * Default implementation to fetch logging constraints remotely from AWS Lambda and S3.
+ */
+class DefaultRemoteLoggingConstraintProvider @JvmOverloads constructor(
+    private val url: URL,
+    private val regionString: String,
+    private val refreshIntervalInSeconds: Int = 1200,
```
Comment (tylerjroach) [inline aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/DefaultRemoteLoggingConstraintProvider.kt:45]: Not going to block on these but, again, val's like these should be const's and better suited to companion object.

**Review by mattcreaser — APPROVED** [review]

**Review by tylerjroach — COMMENTED** [review]
Can you comment on the scenario below. I haven't had a chance to closely inspect.

1. User has sqlcipher db on their device. Key is stored in encrypted shared preferences.
2. User upgrades devices and Shared Preferences are migrated + sqlcipher db.
3. The key is no longer available for database decryption.

Could consider placing the db in noBackupDir? Or if no longer able to decrypt, delete the db and start fresh.

**Review by mattcreaser — APPROVED** [review]
